core/lib/multivm/src/versions/vm_latest/tracers/pubdata_tracer.rs

```diff
@@ -0,0 +1,206 @@
+    // State diffs needed to be published on L1
+    fn get_state_diffs<S: WriteStorage, H: HistoryMode>(
+        state: &ZkSyncVmState<S, H>,
+    ) -> Vec<StateDiffRecord> {
+        sort_storage_access_queries(
+            state
+                .storage
+                .storage_log_queries_after_timestamp(Timestamp(0))
+                .iter()
+                .map(|log| &log.log_query),
+        )
+        .1
+        .into_iter()
+        .filter(|log| log.rw_flag)
+        .filter(|log| log.read_value != log.written_value)
+        .filter(|log| log.address != L1_MESSENGER_ADDRESS)
+        .map(|log| StateDiffRecord {
+            address: log.address,
+            key: log.key,
+            derived_key: log.derive_final_address(),
+            enumeration_index: state
+                .storage
+                .storage
+                .get_ptr()
+                .borrow_mut()
+                .get_enumeration_index(&StorageKey::new(
+                    AccountTreeId::new(log.address),
+                    u256_to_h256(log.key),
+                ))
+                .unwrap_or_default(),
+            initial_value: log.read_value,
+            final_value: log.written_value,
+        })
+        .collect()
+    }
+
+    fn build_pubdata_input<S: WriteStorage, H: HistoryMode>(
+        &self,
+        state: &ZkSyncVmState<S, H>,
+    ) -> PubdataInput {
+        PubdataInput {
+            user_logs: self.get_total_user_logs(state),
+            l2_to_l1_messages: self.get_total_l1_messenger_messages(state),
+            published_bytecodes: self.get_total_published_bytecodes(state),
+            state_diffs: Self::get_state_diffs(state),
+        }
+    }
+}
+
+impl<S, H: HistoryMode> DynTracer<S, H> for PubdataTracer {
+    fn before_execution(
+        &mut self,
+        state: VmLocalStateData<'_>,
+        data: BeforeExecutionData,
+        _memory: &SimpleMemory<H>,
+        _storage: StoragePtr<S>,
+    ) {
+        let hook = VmHook::from_opcode_memory(&state, &data);
+        if let VmHook::PubdataRequested = hook {
+            self.pubdata_info_requested = true;
+        }
+    }
+}
+
+impl<S: WriteStorage, H: HistoryMode> VmTracer<S, H> for PubdataTracer {
+    fn finish_cycle(
+        &mut self,
+        state: &mut ZkSyncVmState<S, H>,
+        bootloader_state: &mut BootloaderState,
+    ) -> TracerExecutionStatus {
+        if !matches!(self.execution_mode, VmExecutionMode::Batch) {
+            // We do not provide the pubdata when executing the block tip or a single transaction
+            if self.pubdata_info_requested {
+                return TracerExecutionStatus::Stop(TracerExecutionStopReason::Finish);
+            } else {
+                return TracerExecutionStatus::Continue;
+            }
+        }
+
+        if self.pubdata_info_requested {
+            // Whenever we are executing the block tip, we want to avoid publishing the full pubdata
```

Comment:
I don't understand how this comment is related to the code below.

Reply:
It was more about the `if` statement, but I'll remove it as it looks self-descripting enough